uploader/push_to_sling.py

```diff
@@ -36,74 +68,114 @@
-        print(r.status_code, r.reason)
-        print()
-    # Otherwise just upload as a regular file
+def resolveOption(parserVal, configKey, default):
+    if parserVal is not None:
+        return parserVal
+    elif config is not None and config[configKey] is not None:
+        return config[configKey]
     else:
-        print(path)
-        logger.debug(url)
-        files = {path.name: (path.name, open(path, 'rb'))}
-        r = requests.post(url, headers=HEADERS, files=files, auth=(args.user, pw))
-        print(r.status_code, r.reason)
-        print()
+        return default
+
+
+server = resolveOption(args.server, 'server', DEFAULT_SERVER)
+repository = resolveOption(args.repository, 'repository', DEFAULT_REPOSITORY)
+links = resolveOption(args.links, 'followlinks', DEFAULT_LINKS)
+
+print('Using server: ' + server)
+print('Using repository: ' + repository)
+print('--------------')
+
+# pathlist = Path(args.directory).glob('**/*')
+titleGlobs = config['titles'] if config is not None else ()
+moduleGlobs = config['modules'] if config is not None else ()
+resourceGlobs = config['resources'] if config is not None else '*'
+
+# for path in pathlist:
+for root, dirs, files in os.walk(args.directory, followlinks=links):
+    for file in files:
+        if file == 'pantheon2.yml':
+            continue
+        logger.debug('root: %s', root)
+        logger.debug('file: %s', file)
+        path = PurePath(root + '/' + file)
+
+        # These distinctions aren't important right now but they set us up for later
+        isTitle = matches(path, titleGlobs, 'title')
+        isModule = matches(path, moduleGlobs, 'module') if not isTitle else False
+        isResource = matches(path, resourceGlobs, 'resource') if not isModule else False
+
+        if isTitle or isModule or isResource:
+            base_name = path.stem
+
+            ppath = path
+            hiddenFolder = False
+            while not ppath == PurePath(args.directory):
+                logger.debug('ppath: %s', str(ppath.stem))
+                if ppath.stem[0] == '.':
+                    hiddenFolder = True
+                    break
+                ppath = ppath.parent
+            if hiddenFolder:
+                logger.info('Skipping %s because it is hidden.', str(path))
+                logger.info('')
+                continue
+
+            # parent directory
+            parent_dir_str = str(path.parent.relative_to(args.directory))
+            if parent_dir_str == '.':
+                parent_dir_str = ''
+            logger.debug('parent_dir_str: %s', parent_dir_str)
+            # file becomes a/file/name (no extension)
+
+            url = server + "/content/repositories/" + repository
+            if parent_dir_str:
+                url += '/' + parent_dir_str
+
+            logger.debug('base name: %s', base_name)
+
+            # Asciidoc content (treat as a module)
+            if path.suffix == '.adoc' or path.suffix == '.asciidoc':
```

Comment:
I think this condition should be replaced with:
```python
if isModule:
```
This way we upload modules as modules.

Reply:
Good catch! I was just setting up `isModule` for future use, but I think you have a good point, we might be able to key off that here. However... I have a few thoughts.
- I'm not sure it's wise to key off of `isModule` alone... the user could easily accidentally write a wildcard glob that flags some .jpg's as modules. (Not that the file extension is much more reliable, but any tomfoolery there is more deliberate).
- What happens if we upload a .adoc file as a resource (not a module)? Probably the system would not play nicely with that in its current state. We'd need to test this pretty thoroughly.

Reply:
So...

- the code computes the `isModule` value for each file. Why not just add to the current condition something like (pseudo code):
```python
isModule = matches(path, moduleGlobs, 'module') AND path.endsWith('.adoc') if not isTitle else False
```
There is no way stop all tomfoolery, but at least we can have some basic intelligence

- If a user uploads an adoc file as a non-module (totally valid use case btw), then it won't list as a module and it won't be previewable. But there is no problem there, as fixing the config file and re-uploading should do the trick.

Reply:
I think I'd prefer not to look at file extension at all, since we're on the topic. Perhaps some author names their files .ascii or something and we're not set up to handle that. Let's just trust the authors to get the globs right - we can always tweak it later if need be. (Sounds like that was your original suggestion anyway)

>     * If a user uploads an adoc file as a non-module (totally valid use case btw), then it won't list as a module and it won't be previewable. But there is no problem there, as fixing the config file and re-uploading should do the trick.

The [possible] problem I'm worried about is in the includeProcessor. Right now it's set up to assuming that it can find the adoc content in jcrNode -> asciidocNode -> jcr:content (I think). Modules fit that pattern, but resources (like .jpgs) do not - there is no asciidoc node for .jpgs, for example. So if we upload a .adoc resource the same way we upload a .jpg, the includeProcessor won't handle it correctly.

Reply:
+1 on not using the extension and trusting users to get it right (or making it very clear why they failed).

But I think our include processor needs to be smart about how to include different types of content (e.g. a `pant:module` vs a `pant:title` vs a regular `nt:file`

Reply:
I just checked and right now it's checking for the existence of an "asciidoc" child node before deciding where to ultimately pull the content from. So it may just work as-is, perhaps not the best implementation, but I think it's good enough for now so I retract my concern.